stack/framework/hal/inc/hwcounter.h

```diff
@@ -0,0 +1,78 @@
+/* OSS-7 - An opensource implementation of the DASH7 Alliance Protocol for ultra
+ * lowpower wireless sensor communication
+ *
+ * Copyright 2018 University of Antwerp
+ * Copyright 2018 Cortus
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.* OSS-7 - An opensource implementation of the DASH7 Alliance Protocol for ultra
+ * lowpower wireless sensor communication
+ *
+ * \file cortus_counter.h
+ *
+ * \author yousra.amrani@cortus.com
+ *
+ */
+
+#ifndef HW_COUNTER_H_
+#define HW_COUNTER_H_
```

Comment:
What's the use of this counter compared to hwtimer? Isn't the functionality basically the same?

Reply:
Hi Aloxy, 
The functionality is indeed the same. However, with the current implementation of hwtimer, we can only select 2 frequencies (1kHz or 32kHz) which give us a granularity quite restricted. 
We are using sensors that require a precision close to the microsecond and Cortus chips have a hardware counter implemented so that's why we introduced this set of APIs. An other approach could be to extend hwtimer APIs.

Reply:
We are open to change the hwtimer API to allow other frequencies. This will prevent duplicate code and will limit the amount of APIs we have to maintain.
Is it possible to change this?

Reply:
Yes, It's possible.

Reply:
To be more specific, we need to get an elapsed time in microsecond which is quite not easy to get as Oss7 is using a millisecond tick basis. Our hwtimer driver implementation is actually based on a 1MHz timer so we could get a microsecond value. Unfortunately handling timer overflow is done in timer.c and, again, it is based on a hwtimer tick... Even by extending hwtimer API, working with microsecond is tricky and unnatural. The introduction of hwcounter API is therefore interesting. 
What do you think?

Reply:
The stack's `timer.c` uses a hw timer (with timer ID 0) on a tick basis indeed, but nothing prevents you to call `hw_timer_init()` again for another timer ID and frequency I think? This would allow you to expose both the existing timer and the counter peripheral using the same API? Or am I misunderstanding?

Reply:
Indeed, we can handle it with a 2nd timer without modifying timer API. Then adding an enum for 1MHz frequency in timer.h should do it.

Reply:
Hi, any update on this?

Reply:
Hello, yes I removed hwcounter driver from this PR so it should be ok.